NumPy 2.0 breaks some NEST tests

NumPy 2.0 has been released on 16 June 2024 with some backward-incompatible changes and expired deprecations ().

This breaks some NEST tests: https://github.com/heplesser/nest-simulator/actions/runs/9538215682/job/26287160722#step:23:9194

The problem is related to changes in NumPy: 
```
FAILED result/share/nest/testsuite/pytests/test_connect_arrays.py::TestConnectArrays::test_connect_arrays_differnt_alpha 
   - AttributeError: `np.string_` was removed in the NumPy 2.0 release. Use `np.bytes_` instead.
```
and related to this code:
```
result/lib/python3.9/site-packages/nest/lib/hl_api_connections.py:254: in Connect
    syn_param_keys = numpy.array(list(reduced_processed_syn_spec.keys()), dtype=numpy.string_)
```

--- a/pynest/nest/lib/hl_api_connections.py
+++ b/pynest/nest/lib/hl_api_connections.py
@@ -252,11 +252,9 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
 
         if len(reduced_processed_syn_spec) > 0:
             syn_param_values = numpy.zeros([len(reduced_processed_syn_spec), len(pre)])
-
             for i, value in enumerate(reduced_processed_syn_spec.values()):
                 syn_param_values[i] = value
         else:
-            syn_param_keys = None
             syn_param_values = None
 
         connect_arrays(pre, post, weights, delays, synapse_model, reduced_processed_syn_spec.keys(), syn_param_values)

--- a/pynest/nest/lib/hl_api_connections.py
+++ b/pynest/nest/lib/hl_api_connections.py
@@ -251,7 +251,6 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
         }
 
         if len(reduced_processed_syn_spec) > 0:
-            syn_param_keys = numpy.array(list(reduced_processed_syn_spec.keys()), dtype=numpy.bytes_)
             syn_param_values = numpy.zeros([len(reduced_processed_syn_spec), len(pre)])
 
             for i, value in enumerate(reduced_processed_syn_spec.values()):
@@ -260,7 +259,7 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
             syn_param_keys = None
             syn_param_values = None
 
-        connect_arrays(pre, post, weights, delays, synapse_model, syn_param_keys, syn_param_values)
+        connect_arrays(pre, post, weights, delays, synapse_model, reduced_processed_syn_spec.keys(), syn_param_values)
 
         return
 

--- a/pynest/nest/lib/hl_api_connections.py
+++ b/pynest/nest/lib/hl_api_connections.py
@@ -251,7 +251,7 @@ def Connect(pre, post, conn_spec=None, syn_spec=None, return_synapsecollection=F
         }
 
         if len(reduced_processed_syn_spec) > 0:
-            syn_param_keys = numpy.array(list(reduced_processed_syn_spec.keys()), dtype=numpy.string_)
+            syn_param_keys = numpy.array(list(reduced_processed_syn_spec.keys()), dtype=numpy.bytes_)
             syn_param_values = numpy.zeros([len(reduced_processed_syn_spec), len(pre)])
 
             for i, value in enumerate(reduced_processed_syn_spec.values()):

--- a/pynest/pynestkernel.pyx
+++ b/pynest/pynestkernel.pyx
@@ -297,7 +297,7 @@ cdef class NESTEngine:
         if delays is not None and  not (isinstance(delays, numpy.ndarray) and delays.ndim == 1):
             raise TypeError('delays must be a 1-dimensional NumPy array')
         if syn_param_keys is not None and not ((isinstance(syn_param_keys, numpy.ndarray) and syn_param_keys.ndim == 1) and
-                                              numpy.issubdtype(syn_param_keys.dtype, numpy.string_)):
+                                              numpy.issubdtype(syn_param_keys.dtype, numpy.bytes_)):
             raise TypeError('syn_param_keys must be a 1-dimensional NumPy array of strings')
         if syn_param_values is not None and not ((isinstance(syn_param_values, numpy.ndarray) and syn_param_values.ndim == 2)):
             raise TypeError('syn_param_values must be a 2-dimensional NumPy array')

--- a/pynest/pynestkernel.pyx
+++ b/pynest/pynestkernel.pyx
@@ -296,19 +296,14 @@ cdef class NESTEngine:
             raise TypeError('weights must be a 1-dimensional NumPy array')
         if delays is not None and  not (isinstance(delays, numpy.ndarray) and delays.ndim == 1):
             raise TypeError('delays must be a 1-dimensional NumPy array')
-        if syn_param_keys is not None and not ((isinstance(syn_param_keys, numpy.ndarray) and syn_param_keys.ndim == 1) and
-                                              numpy.issubdtype(syn_param_keys.dtype, numpy.bytes_)):
-            raise TypeError('syn_param_keys must be a 1-dimensional NumPy array of strings')
         if syn_param_values is not None and not ((isinstance(syn_param_values, numpy.ndarray) and syn_param_values.ndim == 2)):
             raise TypeError('syn_param_values must be a 2-dimensional NumPy array')
 
-        if not len(sources) == len(targets):
+        if len(sources) != len(targets):
             raise ValueError('Sources and targets must be arrays of the same length.')
-        if weights is not None:
-            if not len(sources) == len(weights):
+        if weights is not None and len(sources) != len(weights):
                 raise ValueError('weights must be an array of the same length as sources and targets.')
-        if delays is not None:
-            if not len(sources) == len(delays):
+        if delays is not None and len(sources) != len(delays):
                 raise ValueError('delays must be an array of the same length as sources and targets.')
         if syn_param_values is not None:
             if not len(syn_param_keys) == syn_param_values.shape[0]:
@@ -338,8 +333,8 @@ cdef class NESTEngine:
         # Storing parameter keys in a vector of strings
         cdef vector[string] param_keys_ptr
         if syn_param_keys is not None:
-            for i, key in enumerate(syn_param_keys):
-                param_keys_ptr.push_back(key)
+            for key in syn_param_keys:
+                param_keys_ptr.push_back(key.encode('utf8'))
 
         cdef double[:, ::1] param_values_mv
         cdef double* param_values_ptr = NULL

